From 0d7159ca52771daa8b5cafc422b6c140aa7a11d7 Mon Sep 17 00:00:00 2001
From: Philip Hazel <Philip.Hazel@gmail.com>
Date: Fri, 1 Mar 2024 17:23:34 +0000
Subject: [PATCH] Fix incorrect compiling when variable-length lookbehind's
 first branch was not the shortest and there was a backreference to a
 groupwithin the lookbehind.

Cherry-picked from 4a6a8b056f39079d5e958eac84c2ad173f4680bc.
---
 src/pcre2_compile.c  |  2 +-
 testdata/testinput2  |  4 ++++
 testdata/testoutput2 | 10 ++++++++++
 3 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/pcre2_compile.c b/src/pcre2_compile.c
index 8b364977..71c2dbca 100644
--- a/src/pcre2_compile.c
+++ b/src/pcre2_compile.c
@@ -9908,7 +9908,7 @@ do
   *bptr |= branchlength;  /* branchlength never more than 65535 */
   bptr = *pptrptr;
   }
-while (*bptr == META_ALT);
+while (META_CODE(*bptr) == META_ALT);
 
 /* If any branch is of variable length, the whole lookbehind is of variable
 length. If the maximum length of any branch exceeds the maximum for variable
diff --git a/testdata/testinput2 b/testdata/testinput2
index 97608542..9b4c0684 100644
--- a/testdata/testinput2
+++ b/testdata/testinput2
@@ -6099,4 +6099,8 @@ a)"xI
 /(a(?1)z||(?1)++)$/
     abcd\=disable_recurseloop_check
 
+/(((?<=123?456456|ABC)))(?<=\2)../
+    ABCDEFG
+    12345645678910 
+
 # End of testinput2
diff --git a/testdata/testoutput2 b/testdata/testoutput2
index ee1f70b0..30c5a83e 100644
--- a/testdata/testoutput2
+++ b/testdata/testoutput2
@@ -18038,6 +18038,16 @@ Failed: error -47: match limit exceeded
  0: 
  1: 
 
+/(((?<=123?456456|ABC)))(?<=\2)../
+    ABCDEFG
+ 0: DE
+ 1: 
+ 2: 
+    12345645678910 
+ 0: 78
+ 1: 
+ 2: 
+
 # End of testinput2
 Error -70: PCRE2_ERROR_BADDATA (unknown error number)
 Error -62: bad serialized data
-- 
2.51.2.vfs.0.0

